packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2551,6 +2551,21 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       _stopCursorTimer(resetCharTicks: false);
       _startCursorTimer();
     }
+
+    // Trigger haptic feedback when handle drag changes selection on Android.
+    switch (defaultTargetPlatform) {
+      case TargetPlatform.android:
+        if (!selection.isCollapsed && cause == SelectionChangedCause.drag) {
```

Comment:
Could you explain why the feedback should only be triggered when the selection isn't collapsed?

Reply:
When the selection is collapsed there are some other feedbacks already (for example long-press word selection). This can result in double feedback in some cases.

Reply:
The long-press start feedback for material `TextField` is at https://github.com/flutter/flutter/blob/fd54ddf5cf4ccc920139cb1e4f3beef7a199cd16/packages/flutter/lib/src/material/text_field.dart#L109

However, the feedback for when expanded selection adds or removes a character (but not when the selection is expanded) can't really be handled by the gesture detector callbacks in `_TextFieldSelectionGestureDetectorBuilder`. The gestures that update the selection incrementally are `onSingleLongTapMoveUpdate` but also text selection overlay handle drag. These both trigger `_handleSelectionChanged` in `EditableTextState` when the selected position updates the selection.

This is why it made sense to me to do it where it can be implemented in the simplest way without accessing the `EditableText` state and comparing selection against previous values in multiple places.

Reply:
Performing such checks doesn't seem to be particularly hard? If possible the behavior shouldn't be hard-coded into `EditableText`.

Reply:
@chunhtai do you by chance have the link to the android code that does that? Also I wonder if the same interaction applies to selectable area

Reply:
no i don't have the link to the code, I can look for it though. and yes this applies to selectableregion/selectionarea as well.

Reply:
Yeah chromium implemented that for selection changes: https://source.chromium.org/chromium/chromium/src/+/main:content/public/android/java/src/org/chromium/content/browser/selection/SelectionPopupControllerImpl.java;bpv=1;bpt=1?q=TEXT_HANDLE_MOVE&ss=chromium